lib/Sema/CSApply.cpp

```diff
@@ -5475,22 +5482,31 @@ static void checkNonEphemeralArgumentConversion(
       subExpr = boe->getSubExpr();
       continue;
     }
+
+    // If we found the appropriate conversion expr, note it.
+    if (auto *iotpe = dyn_cast<InOutToPointerExpr>(subExpr)) {
+      subExpr = iotpe->getSubExpr()->getValueProvidingExpr();
+      conversion = ConversionRestrictionKind::InoutToPointer;
+    } else if (auto *atpe = dyn_cast<ArrayToPointerExpr>(subExpr)) {
+      subExpr = atpe->getSubExpr()->getValueProvidingExpr();
+      conversion = ConversionRestrictionKind::ArrayToPointer;
+    } else if (auto *stpe = dyn_cast<StringToPointerExpr>(subExpr)) {
+      subExpr = stpe->getSubExpr()->getValueProvidingExpr();
+      conversion = ConversionRestrictionKind::StringToPointer;
+    }
     break;
   }
 
-  Optional<ConversionRestrictionKind> conversion;
-  if (isa<InOutToPointerExpr>(subExpr))
-    conversion = ConversionRestrictionKind::InoutToPointer;
-  else if (isa<ArrayToPointerExpr>(subExpr))
-    conversion = ConversionRestrictionKind::ArrayToPointer;
-  else if (isa<StringToPointerExpr>(subExpr))
-    conversion = ConversionRestrictionKind::StringToPointer;
-
-  if (conversion && !cs.isConversionNonEphemeral(*conversion, argLocator))
+  if (conversion && !cs.isConversionNonEphemeral(*conversion, argLocator)) {
+    auto lastPathElt = argLocator.last();
+    assert(lastPathElt &&
+           lastPathElt->getKind() == ConstraintLocator::ApplyArgToParam);
+    auto argIdx = lastPathElt->getValue();
     diagnoseIllegalEphemeralConversion(
-        cs.getASTContext(), *conversion, argExpr, argType, param.getPlainType(),
-        callee, fnType, argLocator.getAnchor(),
+        cs.getASTContext(), *conversion, argIdx, subExpr, argType,
```

Comment:
Now that you're passing `subExpr`, I don't think you need to pass `argType`.

Reply:
`diagnoseIllegalEphemeralConversion` will eventually also want to be able to take expressions that haven't had solutions applied to them, so I *believe* the type needs to be passed separately (at least, I'm not sure whether it's safe to rely on CSApply giving us back the same `InOutExpr` such that we can query the type from the constraint system).